### PR TITLE
Docker Container labeling docs

### PR DESCRIPTION
**Pull request by tmjd**
Still need to update calicoctl node run reference page

### CODE REVIEW

**Comment by djosborne** [comment]
Fixes #472

**Comment by djosborne** [comment]
I don't particularly like the name "Simple Profile Policy". Profiles and Policies are two different objects in the Calico data model. I understand that we're using "Policy" here to mean general networking policy. But I wonder if there's a different word we can use so as to not confuse users. 

- Simple Profile Security?
- Simple Security with Profiles?

Open to other suggestions

**Comment by tmjd** [comment]
How about for the 3 different tutorials these names
* Security using Calico Profiles
* Security using Calico Profiles and Policy
* Security using Calico Labels and Policy

**Comment by djosborne** [comment]
I like it a lot. I'd revise the third one though, it's not "Calico Labels". Maybe

- Security using Calico Label-based Policy

?

**Comment by tmjd** [comment]
* Security using Calico Label-based Policy
Could work but I feel like that could also describe the "Profiles and Policy" page too, what about
* Security using Docker Labels and Calico Policy

**Comment by djosborne** [comment]
I suppose they are "Docker Labels", so that sounds good.

File: master/getting-started/docker/index.md [hunk 1]
```diff
@@ -32,15 +32,20 @@ you need to install and use Calico.
...
-#### [Simple Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/simple-policy)
+#### [Simple Profile Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/simple-profile-policy)
 
 The above guide demonstrates Calico connectivity cross host, and how to limit
 that connectivity using simple Calico policy.
 
-#### [Advanced Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/advanced-policy)
+#### [Advanced Profile Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/advanced-profile-policy)
 
 The above guide digs deeper into advanced policy configurations for workloads.
```
Comment (robbrockbank) [inline master/getting-started/docker/index.md]: I think this needs to be expanded now to indicate that it uses a per-network profile to apply per-network sets of labels which can then be used to apply Calico label-based policy.

Perhaps not quite worded like that though :-/

File: master/getting-started/docker/index.md [hunk 2]
```diff
@@ -32,15 +32,20 @@ you need to install and use Calico.
...
 
+#### [Advanced Label Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/advanced-label-policy)
+
+The above guide demonstrates Calico connectivity cross host, and how to limit
```
Comment (robbrockbank) [inline master/getting-started/docker/index.md]: Whereas here, we will be using the container-specific labels to apply Calico label-based policy.  This allows policy tweaks at the container level rather than at the network level.

File: master/getting-started/docker/tutorials/advanced-label-policy.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+explicitly allowed through configuration of the globally defined policy which
+uses selectors to determine which subset of the policy is applied to each
+container based on their labels.  This approach provides a powerful way to
+group together all of your network Policy, makes it easy to reuse policy in
+different networks, and makes it easier to define policy that extends across
+different orchestration systems that use Calico.
+
+## Enabling Docker Networking Container Labels Policy
+
+To enable the Container Labels Policy with Docker Networking the flag
```
Comment (robbrockbank) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: Perhaps should spell out that all calico/node containers need to be started with this option to avoid a label/profile mix

File: master/getting-started/docker/tutorials/advanced-label-policy.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+title: Advanced Container Labels Policy
+---
+
+## Background
+
+With Calico as a Docker network plugin, Calico can be configured to extract the
+labels on a container and apply them to the workload endpoint for use with
+Calico
+[policy]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/policy).
+When Calico is configured to use Container Labels, profiles are not created and
```
Comment (robbrockbank) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: I would probably add the detail about the profile as a separate para after this one - so we can spell out that the default is full isolation between containers until policy is added.

File: master/getting-started/docker/tutorials/advanced-label-policy.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+## Enabling Docker Networking Container Labels Policy
+
+To enable the Container Labels Policy with Docker Networking the flag
+`--use-docker-networking-container-labels` must be passed when starting
+calico/node with the `calicoctl node run` command.
+
+## Managing Calico policy for a network
+
+This section provides an example applying policy using the approach described
+above once Container Labels are enabled.
```
Comment (robbrockbank) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: I don't think container labels should be capitalized.

File: master/getting-started/docker/tutorials/advanced-label-policy.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+-  Frontend containers can only access the Database containers over TCP to port 3306.
+   For now we'll assume no other connectivity is allowed to/from the frontend.
+-  Database containers have no isolation between themselves (to handle synchronization
+   within a cluster).  This could be improved by locking down the port ranges and
+   protocols, but for brevity we'll just allow full access between database
+   containers.
+
+### Global policy applied through label selection
+
+This example demonstrates using global selector-based policy with labels
```
Comment (robbrockbank) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: Do we use the term label-based or selector-based policy?  I forget which one is more widely used - but we should try to be consistent ..

File: master/getting-started/docker/index.md (hunk 1 above)
Comment (tmjd) [inline master/getting-started/docker/index.md]: What did you think about the titles Dan and I were discussing
* Security using Calico Profiles
* Security using Calico Profiles and Policy
* Security using Docker Labels and Calico Policy
I think the detail of what happens with the profiles creating labels can be left as a detail in the tutorial.  Do you think it needs to be in the title itself?

File: master/getting-started/docker/tutorials/advanced-label-policy.md (hunk 6 above)
Comment (tmjd) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: @djosborne do you know which is more widely used?
I'll make the advanced-profile-policy consistent too, this was copied and modified from there.

File: master/getting-started/docker/index.md (hunk 1 above)
Comment (robbrockbank) [inline master/getting-started/docker/index.md]: I think the titles look good.  I agree the bulk of the description can be left to the main article, but we do need a succinct description to lead people to the appropriate article.

Comment (robbrockbank) [inline master/getting-started/docker/index.md]: I also think that brief summary could be added to the initial intro for each of the main docs.

File: master/getting-started/docker/tutorials/advanced-label-policy.md (hunk 6 above)
Comment (djosborne) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: Could have sworn that back when we were adding it, it was "label-based policy". I found a reference to it as such on the trello board too.

Comment (tmjd) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: `grep -Ri 'label[- ]based'` does not appear at all in the docs but 'selector[- ]based' appears 17 times in the master section of the docs.  Should I make the switch?

Comment (djosborne) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: Yep, sounds like selector-based wins

Comment (tmjd) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: @djosborne uh so I'm not making the switch then even though you said yep?

**Comment by tomdee** [comment]
This will need to be merged into the v2.1 dir too as part of this PR.

File: master/getting-started/docker/tutorials/advanced-label-policy.md (hunk 6 above)
Comment (djosborne) [inline master/getting-started/docker/tutorials/advanced-label-policy.md]: Correct, sorry for the confusion

File: master/getting-started/docker/index.md [hunk 7]
```diff
@@ -32,14 +32,27 @@ you need to install and use Calico.
...
-that connectivity using simple Calico policy.
+that connectivity using simple Calico profiles.  One profile is created for
+each network and the connectivity is defined as policy on each profile.
 
-#### [Advanced Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/advanced-policy)
+#### [Security using Calico Profiles and Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/security-using-calico-profiles-and-policy)
 
 The above guide digs deeper into advanced policy configurations for workloads.
+There is stil one profile created for each network but now the profiles define
+labels that are added to each container added to the network.  The policy uses
```
Comment (djosborne) [inline master/getting-started/docker/index.md]: added to -> inherited by

File: master/getting-started/docker/index.md [hunk 8]
```diff
@@ -32,14 +32,27 @@ you need to install and use Calico.
...
+the labels in selectors to configure connectivity.
+
+#### [Security using Docker Labels and Calico Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy)
+
+The above guide demonstrates Calico connectivity cross host, and how to limit
+that connectivity using Calico policy with Docker container's labels.  In this
+walkthrough the profiles are not created for each network but instead the labels
```
Comment (djosborne) [inline master/getting-started/docker/index.md]: > the labels on the Docker containers are added to the workloads.

I don't think this will be clear for a new user. I assume you mean it's "added to the workload endpoint", but even that is a technical detail I don't think needs to be disclosed yet.

The only points I think need to be stressed are:

- Doesn't use profiles at all
- Assigns labels to containers at runtime
- Creates policy which applies based off those labels

The above guide demonstrates Calico connectivity cross host without using Profiles at all. Instead, Calico Policies are defined which apply to containers depending on the labels assigned to them at runtime. This allows for policy adjustment at the container level rather than at the network level.

Comment (djosborne) [inline master/getting-started/docker/index.md]: I think the brief summaries here and the intro paragraph at the beginning of each of these is very important to get right now that we're offering so many different paths and methods for applying policy.

File: master/getting-started/docker/tutorials/security-using-calico-profiles.md [hunk 9]
```diff
@@ -1,8 +1,21 @@
 ---
-title: Simple Policy
+title: Security using Calico Profiles
 ---
 
-The following guide walks through a simple policy demo.
+## Background
+
+With Calico as a Docker network plugin, Calico uses an identically named
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-calico-profiles.md]: Food for thought, a lot of the stuff mentioned in this paragraph can easily be "proven" showing sample output. I wonder if in the future we should rework this guide to focus on how this stuff works from a user perspective, rather than (what this guide seems to focus on which is) cross-host connectivity. 

No action here, just thinking aloud.

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 10]
```diff
@@ -0,0 +1,186 @@
+---
+title: Security using Docker Labels and Calico Policy
+---
+
+## Background
+
+With Calico as a Docker network plugin, Calico can be configured to extract the
+labels on a container and apply them to the workload endpoint for use with Calico
+[policy]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/policy).
+All traffic must be explicitly allowed through configuration of the globally
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: I don't want people to think "All traffic must be explicitly allowed" is a prerequisite or anything. Can you add in "By default, Calico blocks all traffic unless it has been explicitly allowed..."

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+powerful way to group together all of your network Policy, makes it easy to
+reuse policy in different networks, and makes it easier to define policy that
+extends across different orchestration systems that use Calico.
+
+When Calico is configured to use container labels, profiles are not created and 
+have no impact on any container traffic.
+
+## Enabling Docker Networking Container Labels Policy
+
+To enable the container labels Policy with Docker Networking the flag
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: >To enable the container labels Policy with Docker Networking

wording seems off here...

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+`--use-docker-networking-container-labels` must be passed when starting
+calico/node with the `calicoctl node run` command.  All calico/node instances
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: Don't we have a way to set global Calico settings in etcd? I think we have it for felix, not sure if calico/node startup supports it though

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+On any host in your Calico / Docker network, run the following command:
+
+```
+docker network create --driver calico --ipam-driver calico-ipam net1
+```
+
+#### 2. Create the Labeled Workloads
+
+Create the workloads as docker containers with appropriate labels.
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: With this line up here the user might think they're supposed to do something, but the command this is referencing is 3 paragraphs down. This line should probably just be moved right before the bash output below.

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+```
+cat << EOF | calicoctl create -f -
+- apiVersion: v1
+  kind: policy
+  metadata:
+    name: database
+  spec:
+    order: 0
+    selector: role == 'database'
+    ingress:
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: duplicate line

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+   rule to allow all TCP traffic to port 3306 on endpoints that have the label
+   `role = database` (i.e. to database containers)
+
+For details on all of the possible match criteria, see the
+[policy resource]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/profile)
+documentation.
+
+## Multiple networks
+
+Whilst the Docker API supports the ability to attach a container to multiple
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: Maybe we can try to reword this to make it sound more positive?

?While some network providers tend to use different networks to enforce isolation, Calico philosophy opts to put all containers in the same network, and separate them using Calico policy. For this reason, Calico does not support attaching a container to multiple docker networks.

File: master/getting-started/docker/index.md [hunk 16]
```diff
@@ -32,14 +32,25 @@ you need to install and use Calico.
...
+#### [Security using Calico Profiles and Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/security-using-calico-profiles-and-policy)
 
 The above guide digs deeper into advanced policy configurations for workloads.
+There is stil one profile created for each network but now the profiles define
+labels that are inherited by each container added to the network.  The policy uses
+the labels in selectors to configure connectivity.
+
+#### [Security using Docker Labels and Calico Policy]({{site.baseurl}}/{{page.version}}/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy)
+
+The above guide demonstrates Calico connectivity cross host without using
```
Comment (djosborne) [inline master/getting-started/docker/index.md]: This demo doesn't actually demonstrate cross-host connectivity - it's all done on one host (which I think is fine)

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+We set labels on each container indicating the role, in our case frontend
+or database.  The labels are applied directly to each container and must be
+prefixed with `org.projectcalico.label.` for them to be extracted and applied
+to the workload endpoint.
+
+We have decided to use the label `role` indicating the role and a value of
+either `frontend` or `database`.  
+
+Create the workloads as docker containers with appropriate labels.
+```
```
Comment (djosborne) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: need to throw an extra \n here or it jekyll won't render it correctly

**Comment by djosborne** [comment]
Last two comments, then LGTM

File: master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md (hunk 17 above)
Comment (tmjd) [inline master/getting-started/docker/tutorials/security-using-docker-labels-and-calico-policy.md]: I found a few other places that need that too.
